### ISSUE TITLE
Regression testing

### ISSUE THREAD

**Issue by ttuegel** [runtimeverification/haskell-backend#2673]
This is a conversation to brainstorm and evaluate methods to reduce the burdens that regression testing puts on team and technology.

## Needs

The team needs low-latency feedback on potential regressions to performance and correctness in downstream projects.

## Status

We maintain our own regression test suite for minimal, feature-oriented examples. This discussion does not concern these tests.

We also maintain a suite of selected proofs from [kframework/evm-semantics](https://github.com/kframework/evm-semantics] and [kframework/wasm-semantics](https://github.com/kframework/wasm-semantics). To avoid depending on those repositories directly, we collect the output of the frontend and run tests directly on the captured Kore files. The tests are re-captured weekly in an automatically-generated pull request. The process for capturing tests is fragile and the tests are frequently broken. The downstream tests are also not of uniform quality; some proofs are very refined and others not at all refined. 

We automatically run a detailed performance regression analysis on the captured tests, comparing performance with the `master` branch on every pull request.

The current regression testing regime puts almost all of the burden of testing on our team.

## Proposal 1: Push regression testing downstream

Proposal 1 is a social solution.

It is ultimately the responsibility of downstream projects to integrate frequently changes from the backend upstream. This is usually called "maintenance" and projects that don't do it are "unmaintained". It is not the responsibility of this team to maintain unmaintained downstream projects. When there is a regression, downstream projects will recognize that they depended on an "accidental feature" of the backend, or else the regression would have been caught by the backend's own test suite. Downstream project maintainers will be responsible for writing a feature request in this event.

**Pros:**

- Reduced burden on this team.
- Reduced burden on testing infrastructure.
- Increased responsibility of downstream project teams.
- Discovering accidental features makes the system more robust in the end.

**Cons:**

- Increased latency if downstream projects are not diligent.
- Loss of detailed performance regression information.

## Proposal 2: Run regression tests directly from downstream repositories

Proposal 2 is a technical solution.

Regression testing is significantly complicated by the fragile process we have for updating the tests. Instead of capturing the tests, we will run the tests directly in the downstream repositories by (locally, on CI) updating the backend module and going through the usual testing process. This eliminates the technical challenge of updating tests.

The backend cannot include the downstream repositories as a submodule, so for stability we would track the latest commit SHA in a file. Updating the tests is as simple as updating the file.

**Pros:**

- Reduced burden on this team.

**Cons:**

- Increased burden on testing infrastructure.
- The team is still responsible for maintaining accidental features.
- Decreased responsibility of downstream project teams.
- Not simple to run downstream regression tests locally.

**Comment by ttuegel** [comment]
There are mitigations for the disadvantages of Proposal 1, but they require global cooperation with the other teams:

1. We can mitigate the loss of detailed performance information by exporting that feature from our repository into the downstream projects.
2. We can reduce overall latency by addressing sources of latency elsewhere in the system. A large source of latency is updating the frontend repository's backend submodule.
   1. We could reduce latency here by distributing the backend separately, so that every downstream project would have separate submodules for the backend and the frontend.
   2. Alternatively, we could merge the backend repository into [kframework/k](https://github.com/kframework/k) directly, and develop the backend and frontend from the same repository.

**Comment by ehildenb** [comment]
## Proposal 1

This is closer to how the existing setup is. I think the problem that comes up the most often is not diligence, because in our case we are automatically opening PRs into downstream projects and testing them, and those get checked fairly frequently. The biggest problem is competing demands, AFAICT. Some changes which improve the KWasm situation end up breaking KEVM, for example, in a way that means that KEVM is broken until some other fixes make it through the whole backend -> frontend -> semantics cycle.

Even with automatically opening PRs and checking those PRs semi-frequently, some of our dependency chains are long (eg. ERC20 -> KEVM -> Frontend -> Backend). In one current example, KEVM did _not_ catch a performance regression because it just barely made it past the KEVM timeout, but the ERC20 tool _did_ catch a performance regression because it made it much slower. Now the ERC20 tool is in a situation where it cannot update the KEVM submodule until we see a performance increase, but there would not have been a way to catch this situation ahead of time without egregious amounts of speculative testing of downstream projects.

I personally do not want to move K to a monorepo though, maybe in the future when Java and OCaml backend are gone. I also don't think mono-repoing will solve the "competing demands" issue, it just means we'll see more frequent commits to K master, which I think would actually make it _harder_ to deal with competing demands. It could be that _some_ order of the commits coming in allow for a smooth upgrade on both KEVM and KWasm, but that the ordering we actually end up seeing breaks it.

## Proposal 2

My experience with this approach is that (i) it is really slow (takes a lot of CI time), and (ii) it's not very reliable. Especially if there are ever changes which require actually modifying the semantics in order to update the K version it's pegged to, then you need a way to tell it "actually, use my modified branch `b1` on KEVM for testing pr 38 of Haskell backend." Only once the update then makes it to KEVM can you switch back to using `master`.

This is not really a feasible solution for the haskell backend team I don't think, because when a downstream project requires changes to get the update to go through (not the common case, but the most common source of dependency upgrade slowdown), it requires either (i) they know how to fix the downstream projects, or (ii) they ask a downstream project expert to make the changes, and that expert may not have time immediately to do so.

**Comment by ehildenb** [comment]
One thing we could do to shorten latency of the current update process is to have the automerger force-merge update PRs which:

- Are up to date.
- Are passing tests.
- Have not had any modifications to any files in the repo except the submodule.

I think we would see slightly higher merge-rate using this approach, which would help a bit.

I also am personally of the opinion that this problem cannot be perfectly solved without exhaustively testing all downstream projects we care about on every PR into the Haskell backend, and _fixing_ those issues in downstream projects (with draft PRs) _before_ merging the upstream PR. This is too heavy a burden I think though.

When we were developing Firefly, we had a dependency chain Firefly -> KEVM -> blockchain-k-plugin -> K frontend, which caused us lots of problems because frequently changes to blockchain-k-plugin required changes to KEVM. In that case, we required that you post a link in your KEVM (blockchain-k-plugin) PR to the "sister" PR in blockchain-k-plugin (KEVM), with evidence that the two new versions would work together, before either would get approved. That shifts the burden to the PR author to be confident thet problems were not introduced.

**Comment by ehildenb** [comment]
## Proposal 3

Staging branch for Haskell backend.

- New pull requests are opened into `develop` branch, which triggers automatic submodule updates on a few key downstream projects (K, KEVM, KWasm, etc...) also on their `develop` branches.
- Once all the listed downstream projects ("in scope") pass CI on that branch, the backend `master` branch is pushed to be equal to `develop` branch.
- If a change to `develop` branch causes a downstream project to start failing, that change can be forcibly removed from the history, and the PR re-opened with an appropriate regression test which would have caught it manually added to the haskell backend.

This improves on the existing situation slightly, in that it makes it easier to rip out problematic changes before committing fully to them.

**Comment by ttuegel** [comment]
Proposal 3 is the same as Proposal 1, except for changing the names. At present, we have:

- kframework/k submodule -> `<hash>`
- kframework/kore -> `master`

Proposal 3 would change this to:

- kframework/k submodule -> `master`
- kframework/kore -> `develop`

The frontend/backend combination is already tested with downstream projects whenever it is updated, which is practically every commit of the backend anyway. Changing the names doesn't address any of the problems we need to solve.

Disadvantages of Proposal 3 not listed above: 

- The team becomes responsible for providing hotfixes. We don't do this now, so this is creating more busy work.
- Pull requests are based on the `develop` branch, which is subject to having its history rewritten at any time. This creates more integration work for everyone who has work in progress.

**Comment by ehildenb** [comment]
The hotfixes argument is a red-herring. If you don't do hotfixes now I don't think you need to start. I was just providing that as a way to mitigate the problem that "time from PR being opened to it being available in `master` gets longer". Yes it is more integration work, but if your goal is to make sure that changes don't get merged which break downstream projects, you will have to take on more integration work no matter what.

One scenario which we've seen that neither solution can fully address is as follows: Suppose KEVM needs to make a change (called patch `P1`, which may not work on `master` KEVM) to work with the newest Haskell backend. In this case, your need to update your kevm_release tag in haskell-backend to point to something that is not behind `master`, and leave it there until that patch `P1` is merged into KEVM _after_ or _while_ the submodule updates go through. We have no real way to guarantee that this merge will happen quickly, as we don't know all the problems we'll face in updating K and then KEVM. Even if we test haskell backend with K and KEVM (their full test-suites), someone else could merge something into K or KEVM in the multiple hours running those test-suites takes (invalidating the results). So we will be in a situation where the haskell backend points at a non-`master` commit of KEVM for perhaps indefinite periods of time, which I see as a no-go.

Having the staging `develop` branch makes it possible to go through the normal review/testing process for getting changes approved (into `develop`), and then work on locking in a subset of those changes that behave well for all the downstream projects. Hopefully it frequently means "all the changes that have accumulated", but maybe it means one needs to be excluded (causes a performance regression downstream). Note that we could, for example, run KEVM CI on _every_ PR into haskell backend, but KEVM CI takes 2-3 hours. If you'll only end up running selected proofs anyway, you might as well just add those to the Haskell-backend regression test-suite. If you want to fully run the frontend and semantics integration tests and collect performance numbers (and use those numbers to determine merge-ability), I think having a staging branch for those integration tests makes sense purely from a "limiting how much work the CI servers have to do" perspective. Once the downstream integration test suite passes on a commit of `develop`, we can just push `master` directly up to `develop` automatically (no more testing needed), and have high confidence that downstream projects will be able to update quickly (but without having to run downstream CI on every PR).

Either way I think we have a trade-off with how development goes. Either you can get changes in really quick (hotfixes are easy) for downstream discovered bugs, but you risk breaking integration with another downstream project. Or you take longer to curate the batches of changes (downstream projects rarely break), but you can't get any particular change in as fast. We can test and test and test, but unless we make every other repo at the company synchronize on updates to haskell-backend (stop all other merges until haskell backend updates go fully through), we can't guarantee we won't have breakages.

**Comment by ehildenb** [comment]
Anyway, let me know if you need a feed of downstream CI commit release tags setup from the devops repo, and which you would be interested in having.

**Comment by ttuegel** [comment]
> We can test and test and test, but unless we make every other repo at the company synchronize on updates to haskell-backend (stop all other merges until haskell backend updates go fully through), we can't guarantee we won't have breakages.

Just for the purpose of focusing the discussion, I want to remark that preventing breakage is not one of the goals of this discussion or of our current testing regime. We don't have regressions in downstream projects because we they have mandatory testing policies: updates aren't merged until the tests pass. What we have are potential regressions, and the concern here is lowering the latency of the team's feedback loop. You say yourself that batching changes increases latency:

> Or you take longer to curate the batches of changes (downstream projects rarely break), but you can't get any particular change in as fast.

So, from my perspective, doing more batching is a non-starter. Actually, it highlights to me that we are already batching changes, and that might be counterproductive.

Actually, now that I think of it, I should say, I disagree with this:

> I also don't think mono-repoing will solve the "competing demands" issue, it just means we'll see more frequent commits to K master, which I think would actually make it _harder_ to deal with competing demands.

Batching backend updates (as we do now) makes it harder to deal with competing demands because each update is more likely to break the downstream project and it takes greater effort to disentangle the broken change from its neighbors. But as I think more on it, removing this source of batching (the submodule) won't change anything as long as there is a downstream source of batching (the release job).

**Comment by ttuegel** [comment]
> One scenario which we've seen that neither solution can fully address is as follows: Suppose KEVM needs to make a change (called patch `P1`, which may not work on `master` KEVM) to work with the newest Haskell backend. In this case, your need to update your kevm_release tag in haskell-backend to point to something that is not behind `master`, and leave it there until that patch `P1` is merged into KEVM _after_ or _while_ the submodule updates go through. We have no real way to guarantee that this merge will happen quickly, as we don't know all the problems we'll face in updating K and then KEVM. Even if we test haskell backend with K and KEVM (their full test-suites), someone else could merge something into K or KEVM in the multiple hours running those test-suites takes (invalidating the results). So we will be in a situation where the haskell backend points at a non-`master` commit of KEVM for perhaps indefinite periods of time, which I see as a no-go.

This entire scenario is trivially avoided with [feature flags](https://www.martinfowler.com/articles/feature-toggles.html).
